platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/BusyTime.java

```diff
@@ -0,0 +1,143 @@
+     */
+    public BusyTime(final Metrics metrics, final DefaultMetricConfig config) {
+        this(metrics, config, OSTime.getInstance());
+    }
+
+    /**
+     * A constructor where a custom {@link Time} instance could be supplied
+     *
+     * @param config
+     * 		the configuration for this metric
+     * @param time
+     * 		provides the current time
+     */
+    public BusyTime(final Metrics metrics, final DefaultMetricConfig config, final Time time) {
+        this.time = new IntegerEpochTime(time);
+        this.accumulator = metrics.getOrCreate(new IntegerPairAccumulator.Config<>(
+                        config.getCategory(), config.getName(), Double.class, this::busyFraction)
+                .withDescription(config.getDescription())
+                .withUnit("fraction")
+                .withFormat(FloatFormats.FORMAT_1_3)
+                .withCombinedAccumulator(this::statusUpdate)
+                .withLeftInitializer(this.time::getMicroTime)
+                .withRightReset(this::resetStatus));
+    }
+
+    /**
+     * Notifies the metric that work has started
+     */
+    public void startingWork() {
+        accumulator.update(time.getMicroTime(), WORK_START);
+    }
+
+    /**
+     * Notifies the metric that work has finished
+     */
+    public void finishedWork() {
+        accumulator.update(time.getMicroTime(), WORK_END);
+    }
+
+    /**
+     * @return the fraction of time that the thread has been busy
+     */
+    public double getBusyFraction() {
+        return accumulator.get();
+    }
+
+    /**
+     * @return the accumulator that is used to update the metric
+     */
+    public IntegerPairAccumulator<Double> getAccumulator() {
+        return accumulator;
+    }
+
+    private double busyFraction(final int measurementStart, final int status) {
+        final int elapsedTime = time.microsElapsed(measurementStart);
+        if (elapsedTime == 0) {
+            return 0;
+        }
+        final int busyTime;
+        if (isIdle(status)) {
+            busyTime = Math.abs(status) - 1;
+        } else {
+            busyTime = elapsedTime - (status - 1);
+        }
+        return ((double) busyTime) / elapsedTime;
+    }
+
+    private boolean isIdle(final int status) {
+        return status < 0;
+    }
+
+    private long statusUpdate(final long previousPair, final long suppliedValues) {
+        final int measurementStart = IntPairUtils.extractLeft(previousPair);
+        final int currentStatus = IntPairUtils.extractRight(previousPair);
+        final int currentTime = IntPairUtils.extractLeft(suppliedValues);
+        final int statusChange = IntPairUtils.extractRight(suppliedValues);
+
+        if ((statusChange == WORK_START && !isIdle(currentStatus))
+                || (statusChange == WORK_END && isIdle(currentStatus))) {
+            // this means that the metric has not been updated correctly, we will not change the value
+            return previousPair;
```

Comment:
I'm not sure....
If we do log it, there is a very high possibility that we will be spamming the log a lot. If we don't do anything, like now, it will be a silent failure, which is not good. Got any suggestions?

Reply:
If we're spamming the log at high rates because there is a high rate of error, then the metric is bunk.  What contributes to the meaningfulness of the metric is that we are accurately reporting boundaries between busy and idle time periods.   Either we log the error, or we have a meta--trust-metric, a speedometer, that tells us how much misreporting is going on.

Reply:
> When do these conditions happen? Someone using the metric and calling startWork successively or finishWork successively? This would seem to be a fatal logic error on the part of the caller. Would it be appropriate to throw an IllegalStateException to force them to recode their logic?

That could cause the system to die because its not using a metric correctly, I don't think that is desirable behaviour

Reply:
Nothing should crash or stop because a metric is not working properly.

How about logging and using the log rate limiter?

Reply:
Would a `ThresholdLimitingHandler` help?

Reply:
I feel there is something more fundamental to consider.   The only calling methods are `startingWork` and `finishedWork` which should be called on the same thread of execution before and after a runnable or callable executes.  I presume there is exactly 1 BusyTime metric per queue/thread and that they cannot cross call each other's BusyTime metric.   The only sequence algorithmically possible should be `startingWork`, `finishedWork`, `startingWork`, `finishedWork`, etc. in alternating fashion.  The throw of an IllegalStateException would never happen unless something was really wrong.  Which is why I think throwing it is the right behavior, it should never happen in code that is solid enough to make it to production.

Reply:
i implemented a log with a rate limiter